DigitalLearningSolutions.Data/Services/GroupsService.cs

```diff
@@ -0,0 +1,234 @@
+                }
+
+                var candidateProgressOnCourse =
+                    progressDataService.GetDelegateProgressForCourse(oldDetails.Id, groupCourse.CustomisationId);
+                var existingRecordToUpdate = candidateProgressOnCourse.FirstOrDefault(
+                    p => !p.SystemRefreshed && p.Completed == null && p.RemovedDate == null
+                );
+
+                if (existingRecordToUpdate != null)
+                {
+                    var updatedSupervisorAdminId = groupCourse.SupervisorAdminId > 0
+                        ? groupCourse.SupervisorAdminId.Value
+                        : existingRecordToUpdate.SupervisorAdminId;
+                    progressDataService.UpdateProgressSupervisorAndCompleteByDate(
+                        existingRecordToUpdate.ProgressId,
+                        updatedSupervisorAdminId,
+                        completeByDate
+                    );
+                }
+                else
+                {
+                    var newProgressId = progressDataService.CreateNewDelegateProgress(
+                        oldDetails.Id,
+                        groupCourse.CustomisationId,
+                        groupCourse.CurrentVersion,
+                        clockService.UtcNow,
+                        3,
+                        addedByAdminId,
+                        completeByDate,
+                        groupCourse.SupervisorAdminId ?? 0
+                    );
+
+                    var tutorialsForCourse =
+                        tutorialContentDataService.GetTutorialsForCourse(groupCourse.CustomisationId);
+
+                    foreach (var tutorial in tutorialsForCourse)
+                    {
+                        progressDataService.CreateNewAspProgress(tutorial, newProgressId);
+                    }
+                }
+
+                if (newDetails.Email != null)
+                {
+                    var fullName = newDetails.FirstName + " " + newDetails.Surname;
+                    var email = BuildEnrolmentEmail(newDetails.Email, fullName, groupCourse, baseUrl, completeByDate);
+                    emailService.ScheduleEmail(email, AddedByProcess);
+                }
+            }
+        }
+
+        private IEnumerable<Group> GetSynchronisedGroupsForCentre(int centreId)
+        {
+            return groupsDataService.GetGroupsForCentre(centreId)
+                .Where(g => g.ChangesToRegistrationDetailsShouldChangeGroupMembership);
+        }
+
+        private void RemoveDelegateFromGroup(int delegateId, int groupId)
+        {
+            groupsDataService.DeleteGroupDelegatesRecordForDelegate(groupId, delegateId);
+            groupsDataService.RemoveRelatedProgressRecordsForGroupDelegate(groupId, delegateId, clockService.UtcNow);
+        }
+
+        private Email BuildEnrolmentEmail(
+            string emailAddress,
+            string fullName,
+            GroupCourse course,
+            string baseUrl,

```

Comment:
`baseUrl` gets passed down about 5 layers to get here, but comes from a static method on ConfigHelper. Is there a reason not to make the call directly in this method?

Reply:
ConfigHelper and the configuration data is defined in the Web project, which references the Data project. We can't reference the Web project from the Data project or else we start getting cyclical references. We have the same issue with setting up some emails (the ForgotPassword email comes to mind).

Reply:
Ah. That's frustrating. We should change that, but I'll put it in a tech debt ticket.
For now, could you just grab the configuration directly?
You can
- inject it into the service in the constructor as `IConfiguration`
- create a `ConfigHelper` class in the Data project (which will eventually become the only one) and copy the `GetAppRootPath()` method
- get the base url where you need it by using the new helper method.

You'll need to add a reference to Microsoft.Extensions.Configuration or somesuch.
It's possible having another class called `ConfigHelper` will cause problems. If so, I think just give it a different name. Maybe `ConfigurationHelper`.

Could you also stick an `Obsolete` attribute on the existing config helper with a description saying that we're moving to ConfigurationHelper, and that people should copy over the methods they need?